File: Windows/Type4Me/ASR/Providers/CustomASRConfig.cs
using Type4Me.Localization;

namespace Type4Me.ASR.Providers;

public sealed class CustomASRConfig
{
    public string EndpointURL { get; }
    public string? ApiKey { get; }
    public string? SecretKey { get; }
    public string? AppId { get; }
    public string? Region { get; }

    public static ASRProvider Provider => ASRProvider.Custom;
    public static string DisplayName => Loc.L("自定义", "Custom");

    public static CredentialField[] CredentialFields =>
    [
        new() { Key = "endpointURL", Label = "Endpoint URL", Placeholder = Loc.L("wss://... 或 https://...", "wss://... or https://...") },
        new() { Key = "apiKey", Label = "API Key / Token", Placeholder = Loc.L("认证密钥", "Auth key"), IsSecure = true, IsOptional = true },
        new() { Key = "secretKey", Label = "Secret Key", Placeholder = Loc.L("可选", "Optional"), IsSecure = true, IsOptional = true },
        new() { Key = "appId", Label = "App ID", Placeholder = Loc.L("可选", "Optional"), IsOptional = true },
        new() { Key = "region", Label = "Region", Placeholder = Loc.L("可选", "Optional"), IsOptional = true },
    ];

    private CustomASRConfig(string endpointURL, string? apiKey, string? secretKey, string? appId, string? region)
    {
        EndpointURL = endpointURL; ApiKey = apiKey; SecretKey = secretKey; AppId = appId; Region = region;
    }

    public static CustomASRConfig? TryCreate(Dictionary<string, string> credentials)
    {
        if (!credentials.TryGetValue("endpointURL", out var url) || string.IsNullOrEmpty(url)) return null;
        return new CustomASRConfig(url,
            credentials.GetValueOrDefault("apiKey"),
            credentials.GetValueOrDefault("secretKey"),
            credentials.GetValueOrDefault("appId"),
            credentials.GetValueOrDefault("region"));
    }

    public Dictionary<string, string> ToCredentials()
    {
        var d = new Dictionary<string, string> { ["endpointURL"] = EndpointURL };
        if (!string.IsNullOrEmpty(ApiKey)) d["apiKey"] = ApiKey;
        if (!string.IsNullOrEmpty(SecretKey)) d["secretKey"] = SecretKey;
        if (!string.IsNullOrEmpty(AppId)) d["appId"] = AppId;
        if (!string.IsNullOrEmpty(Region)) d["region"] = Region;
        return d;
    }

    public bool IsValid => !string.IsNullOrEmpty(EndpointURL);
}

File: MacOS/Type4Me/ASR/Providers/CustomASRConfig.swift
import Foundation

struct CustomASRConfig: ASRProviderConfig, Sendable {

    static let provider = ASRProvider.custom
    static var displayName: String { L("自定义", "Custom") }

    static var credentialFields: [CredentialField] {[
        CredentialField(key: "endpointURL", label: "Endpoint URL", placeholder: L("wss://... 或 https://...", "wss://... or https://..."), isSecure: false, isOptional: false, defaultValue: ""),
        CredentialField(key: "apiKey", label: "API Key / Token", placeholder: L("认证密钥", "Auth key"), isSecure: true, isOptional: true, defaultValue: ""),
        CredentialField(key: "secretKey", label: "Secret Key", placeholder: L("可选", "Optional"), isSecure: true, isOptional: true, defaultValue: ""),
        CredentialField(key: "appId", label: "App ID", placeholder: L("可选", "Optional"), isSecure: false, isOptional: true, defaultValue: ""),
        CredentialField(key: "region", label: "Region", placeholder: L("可选", "Optional"), isSecure: false, isOptional: true, defaultValue: ""),
    ]}

    let endpointURL: String
    let apiKey: String?
    let secretKey: String?
    let appId: String?
    let region: String?

    init?(credentials: [String: String]) {
        guard let url = credentials["endpointURL"], !url.isEmpty else { return nil }
        self.endpointURL = url
        self.apiKey = credentials["apiKey"]
        self.secretKey = credentials["secretKey"]
        self.appId = credentials["appId"]
        self.region = credentials["region"]
    }

    func toCredentials() -> [String: String] {
        var result = ["endpointURL": endpointURL]
        if let v = apiKey, !v.isEmpty { result["apiKey"] = v }
        if let v = secretKey, !v.isEmpty { result["secretKey"] = v }
        if let v = appId, !v.isEmpty { result["appId"] = v }
        if let v = region, !v.isEmpty { result["region"] = v }
        return result
    }

    var isValid: Bool { !endpointURL.isEmpty }
}
